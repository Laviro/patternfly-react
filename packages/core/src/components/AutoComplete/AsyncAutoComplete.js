import React from 'react';
import { AsyncTypeahead } from 'react-bootstrap-typeahead';
import PropTypes from 'prop-types';

class AsyncAutoComplete extends React.Component {
  state = {
    options: this.props.options,
    isLoading: this.props.isLoading
  };

  onSearchStart = () => this.setState({ isLoading: true });

  onSearchEnd = options => this.setState({ options, isLoading: false });

  // query is the text string entered by the user.
  handleSearch = query => {
    this.onSearchStart();
    Promise.resolve(this.props.onSearch(query)).then(options =>
      this.onSearchEnd(options)
    );
  };

  render = () => (
    <AsyncTypeahead
      {...this.props}
      onSearch={this.handleSearch}
      options={this.state.options}
      isLoading={this.state.isLoading}
    />
  );
}

AsyncAutoComplete.propTypes = {
  onSearch: PropTypes.func.isRequired,
  options: PropTypes.array,
  isLoading: PropTypes.bool,
  delay: PropTypes.number,
  promptText: PropTypes.object,
  searchText: PropTypes.object,
  useCache: PropTypes.bool
};

AsyncAutoComplete.defaultProps = {
  options: [],
  isLoading: false,
  delay: 200,
  promptText: <span>Type to search...</span>,
  searchText: <span>Searching...</span>,
  useCache: true
};

export default AsyncAutoComplete;
