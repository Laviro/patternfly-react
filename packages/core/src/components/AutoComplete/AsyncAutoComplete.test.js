import React from 'react';
import { shallow } from 'enzyme';
import AsyncAutoComplete from './AsyncAutoComplete';

test('AsyncAutoComplete is working !!', () => {
  const optionsArray = ['One', 'Two', 'Three'];
  const timeout = 500;
  const handleSearch = () => {
    setTimeout(() => optionsArray, timeout);
  };
  const component = shallow(
    <AsyncAutoComplete
      useCache
      labelKey="login"
      minLength={0}
      placeholder="Search someone who have forked Patternfly-react.."
      onSearch={handleSearch}
      renderMenuItemChildren={option => <span>option</span>}
    />
  );

  expect(component).toMatchSnapshot();
  component.instance().handleSearch();
  setTimeout(() => {
    component.update();
    expect(component.state().options).toEqual(optionsArray);
    expect(component.state().isLoading).toEqual(false);
  }, timeout + 1000);
});
