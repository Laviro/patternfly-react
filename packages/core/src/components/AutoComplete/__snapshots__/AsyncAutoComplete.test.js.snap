// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncAutoComplete is working !! 1`] = `
<AsyncContainer(OnClickOutside(TypeaheadContainer(WrappedTypeahead)))
  delay={200}
  isLoading={false}
  labelKey="login"
  minLength={0}
  onSearch={[Function]}
  options={Array []}
  placeholder="Search someone who have forked Patternfly-react.."
  promptText={
    <span>
      Type to search...
    </span>
  }
  renderMenuItemChildren={[Function]}
  searchText={
    <span>
      Searching...
    </span>
  }
  useCache={true}
/>
`;
